### ISSUE TITLE
Kani encodes string slices as null-terminated C slices

### ISSUE THREAD

**Issue by giltho** [model-checking/kani#1665]
In Kani's regression test suite, in [offset_u8_ok.rs](https://github.com/model-checking/kani/blob/main/tests/kani/Intrinsics/Offset/offset_u8_ok.rs#L21), an out-of-bound pointer is dereferenced, but not detected by Kani. Here's a reduced code
```rust
#![feature(core_intrinsics)]
use std::intrinsics::offset;

#[kani::proof]
fn test_offset() {
    let s: &str = "123";
    let ptr: *const u8 = s.as_ptr();
    unsafe {
        let x = *offset(ptr, 3);
    }
}
```

The pointer is 1-passed the bound of the object, so it is "weak_valid", in the sense defined by [CompCert](https://compcert.org/doc/html/compcert.common.Values.html#Val.COMPARISONS.weak_valid_ptr), meaning that it is either in-bound or one-passed-the bound of the object. According to C standard, that is not undefined behaviour, and shouldn't be detected as such. However, it seems that, in addition to check that the pointer can be created, this test also checks that it can be dereferenced which is always UB.

@danielsn mentioned it might be on purpose because the stdlib does that sometimes, but I haven't been able to find an example

**Comment by zhassan-aws** [comment]
Interestingly, Kani reports the dereference failure if I change it to an array:
```rust
#![feature(core_intrinsics)]
use std::intrinsics::offset;

#[kani::proof]
fn test_offset() {
    let s = [1u8, 2, 3];
    let ptr: *const u8 = s.as_ptr();
    unsafe {
        let x = *offset(ptr, 3);
    }
}
```

```
$ kani iss1665.rs
<snip>
SUMMARY:
 ** 1 of 8 failed
Failed Checks: dereference failure: pointer outside object bounds
 File: "/home/ubuntu/examples/iss1665.rs", line 9, in test_offset

VERIFICATION:- FAILED

**Comment by giltho** [comment]
This is due to the fact that Kani encodes rust str constants into string constants for CBMC
However, CBMC automatically null-terminates C strings, and therefore, this is indeed not out-of-bound access for CBMC which is why it's not detected.

The solution might be to just have CBMC make non-null terminated constants
